subsys/net/lib/sockets/sockets.c

```diff
@@ -1236,13 +1271,21 @@ int z_impl_zsock_poll(struct zsock_pollfd *fds, int nfds, int poll_timeout)
 			 * and non-offloaded sockets, the offloaded poll handler
 			 * shall return an error.
 			 */
-			return z_fdtable_call_ioctl(vtable, ctx,
-						    ZFD_IOCTL_POLL_OFFLOAD,
-						    fds, nfds, poll_timeout);
+			result = z_fdtable_call_ioctl(vtable, ctx,
+						      ZFD_IOCTL_POLL_OFFLOAD,
+						      fds, nfds, poll_timeout);
+			sys_mutex_unlock(lock);
```

Comment:
I think that mutex should be unlocked before the actual call to the offloaded `poll`. Otherwise, the socket would be locked throughout the `poll` call (and it would only be the first socket on the `fds` list locked which makes little sense).

The offloaded `poll` will still have access to the mutexes, it can handle the synchronization if needed.